pkg/cache/keyfunc.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/tools/cache"
+)
+
+// DeletionHandlingMetaClusterNamespaceKeyFunc checks for
+// DeletedFinalStateUnknown objects before calling
+// MetaClusterNamespaceKeyFunc.
+func DeletionHandlingMetaClusterNamespaceKeyFunc(obj interface{}) (string, error) {
```

Comment:
Should we fold this into MetaClusterNamespaceKeyFunc instead of having this as a separate function?

Reply:
I was thinking in terms of how to best refactor this from an AST perspective, and while in theory it seems like collapsing the `DeletionHandling` one would be simpler and always correct, it is subtly different and I figured we could be more confident in a direct translation to cluster-aware versions of the key funcs instead.

Reply:
For instance, a controller that was *not* using the `DeletionHandlingMetaNamespaceKeyFunc()` would error on tombstones, and likely drop the event from the queue. If we collapsed the definition, their controller would begin handling object deletion in their queues, which they are likely not going to support in their reconciliation code?

Reply:
Good point, but you typically only need to deal with tombstone object content in either a FilterFunc or an event handler func, where you want to evaluate the object to determine if you want to enqueue or not. Once the key is in the queue, standard lister get behavior applies (i.e. you get a not found error).

Reply:
Am fine to defer consolidation as a follow-up if it makes senes